bugsnag-android-core/src/main/java/com/bugsnag/android/internal/JournaledDocument.kt

```diff
@@ -27,20 +33,23 @@ class JournaledDocument
  *                    journal is reloaded)
  * @param version The journal type version (to support migration)
  * @param bufferSize The size of the shared memory buffer (and thus the maximum size of the journal)
- * @param initialDocument The initial document contents (will be shallow copied)
+ * @param highWater The "high water" point. If the size of the serialized journal exceeds this, the
+ *                  journal will be snapshotted when snapshotIfHighWater() is called.
+ * @param initialDocument The initial document contents (will be deep copied)
  */
 constructor(
     baseDocumentPath: File,
     journalType: String,
     version: Int,
     bufferSize: Long,
-    initialDocument: MutableMap<String, Any>
+    private val highWater: Long,
+    initialDocument: Map<String, Any>
 ) : Map<String, Any>, Closeable {
     private val journalPath = getJournalPath(baseDocumentPath)
     private val snapshotPath = getSnapshotPath(baseDocumentPath)
     private val newSnapshotPath = getNewSnapshotPath(baseDocumentPath)
 
-    private val document = initialDocument.toMutableMap()
+    private val document = convertDocumentToConcurrent(initialDocument)
```

Comment:
Is it possible to just expect callers to supply `ConcurrentHashMap`, or will the anticipated usage be passing in non-trivial data here? This would substantially reduce the complexity if this could be mandated as it'd allow for `convertDocumentToConcurrent` and associated functions to be removed.

Reply:
I wanted to ensure that there can be no modify-after-commit, and that requires a deep copy of the structure. I've had too many bad experiences in the past of phantom data changes because the same map got passed into two objects that didn't deep copy, and they then proceeded to stomp all over each others' data.

You can solve this with immutable containers, but this document is designed specifically to be modified (a lot), so I think it would be good to ensure that phantom modifications simply cannot occur by design.

We also won't be able to control the data being added by end users (if they add map metadata, we couldn't ensure the internal data's integrity should they modify the source map elsewhere), but this would be mitigated by deep copying.

Reply:
Probably not strictly a problem: the mutable maps are leaked via delegation. So this is still technically a problem:
```kotlin
(journalDocument["key"] as MutableMap<String, Any?>)["something"] = true
```
I doubt it's worth addressing in this PR though.

Reply:
Yes, there's a big fat warning at the top of the class that says not to modify the document except via the journal commands. There's no way to actually enforce this invariant, unfortunately. But considering this class is not meant to be exposed for the user, we just need to be disciplined about it.